### PR TITLE
[MLIR][TORCH] Add OnnxToTorch lowering for ReduceL1 Op

### PR DESCRIPTION
**Pull request by vinayakdsci**
Adds OnnxToTorch Lowering for the ReduceL1 op.

### CODE REVIEW

**Comment by vinayakdsci** [comment]
cc @vivekkhandelwal1 @renxida @rsuderman @zjgarvey

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:863]: This should be changed to use "operand" instead of "data", since you do not want to include the absolute value op in the noop case.

**Review by zjgarvey — CHANGES_REQUESTED** [review]
Looks pretty good, but should edit the noop case to not include absolute value.

Do you know what is causing the other e2e tests for this op to fail?

**Comment by vinayakdsci** [comment]
> Looks pretty good, but should edit the noop case to not include absolute value.

Thanks! I will fix this and re-push.

> Do you know what is causing the other e2e tests for this op to fail?

The other e2e test `ReduceL1NormWithDTypeModule_basic` fails on my build with `Found value Double but Expected Float`. I tried putting in Dtype conversions to check, but the test fails regardless. Do you have any suspicion on where the test could be failing? If I change dtype from `float64` to `float32` in `build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/test_suite/reduction.py` where the traceback points, I am able to get the test to pass.

**Comment by vinayakdsci** [comment]
Is it possible that there could be a mistake in `projects/pt1/python/torch_mlir_e2e_test/onnx_backends/reductions.py`?

**Comment by zjgarvey** [comment]
> Is it possible that there could be a mistake in `projects/pt1/python/torch_mlir_e2e_test/onnx_backends/reductions.py`?

It is certainly possible. I'll take a look. We should also have someone like @rsuderman or @renxida  start the workflows since I don't think I have permissions to start those. Feel free to resolve the earlier conversations, since you made appropriate changes.

**Review by zjgarvey — APPROVED** [review]
Hi, I think it is actually an issue with either the onnx import or onnx export not respecting the other dtypes. So I wouldn't worry about the other tests with onnx config. 

So I'd say this looks good to me. I won't be able to merge the PR, so please reach out to someone with merge permissions. @vivekkhandelwal1 is very familiar with these reduce ops, so he would be a good one to take a look.

**Comment by vinayakdsci** [comment]
@zjgarvey Thanks for the review! I am waiting for a review from @vivekkhandelwal1 too, in the meantime.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:776]: I think if you just pass operand.getType(), it should suffice

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:779]: If this is the case then the better approach is to extract this code and put it inside a function so that it can be reused for both these lowering and possible future lowerings.

File: projects/pt1/e2e_testing/xfail_sets.py
Comment (vivekkhandelwal1) [inline projects/pt1/e2e_testing/xfail_sets.py:2039]: You should try to fix the remaining tests as well.

File: test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir
Comment (vivekkhandelwal1) [inline test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir:930]: You can remove this test since it's similar to the test above.

File: test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir
Comment (vivekkhandelwal1) [inline test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir:949]: I am not sure if we have added the e2e support for handling the `0-dim` case as mentioned here: https://onnx.ai/onnx/operators/onnx__ReduceL1.html#summary

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vinayakdsci) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:779]: I extracted it into a common implementation, but it seems a little too complex. Is it OK if I push that for review?

File: test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir
Comment (vinayakdsci) [reply test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir:930]: Understood, Will do.

File: test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir
Comment (vinayakdsci) [reply test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir:949]: Should I remove this test then?

File: projects/pt1/e2e_testing/xfail_sets.py
Comment (vinayakdsci) [reply projects/pt1/e2e_testing/xfail_sets.py:2039]: This is where I am running into a problem, as `ReduceL1NormWithDTypeModule_basic` is failing with `Found dtype Double but expected Float`. No matter In which Order I apply type conversion, the error is not going away. Do you have any suspicion about what could be going on?

File: test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir
Comment (vivekkhandelwal1) [reply test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir]: Yeah, you can remove this test. But, I think it's necessary to error out on this as well.

Comment (vinayakdsci) [reply test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir]: So I should add a check in e2e tests?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:156]: Why do we need this functionality?

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vinayakdsci) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:156]: This is to replace binder.op when the ReduceSum op is the last operation performed on the data. If it is an intermediate operation, and its result will be passed into another operation(imagine ReduceL2) We might need to create a new op value, as a replacement will give a segfault here.
That is why I put in  this check, just for future proofing.

Comment (vinayakdsci) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:156]: I am sure it can be leveraged in the implementation of ReduceL2 too.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vinayakdsci) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:50]: But wouldn't this make it necessary to initialize a variable to create a reference even when it is not required?

Comment (vinayakdsci) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:50]: Right now I can pass in  a `nullptr`.

Comment (vinayakdsci) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:50]: No, sorry, I was wrong here. When it is not required, I can simply pass in  a reference to operand, as it won't be modified anyway.

**Comment by vinayakdsci** [comment]
@vivekkhandelwal1 could `ReduceL1NormComplexModule_basic` be failing because of failure in import/export tests for `LinalgVectorNorm`? `LinalgVectorNormComplexModule_basic` is failing with the same error message when run.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:138]: Instead, you can pass `originalOperand` as an arg corresponding to `storeResult` that way you won't have a need for an extra arg. For this case you can just use `storeResult` and for the other ones you can modify it.

File: projects/pt1/e2e_testing/xfail_sets.py
Comment (vivekkhandelwal1) [reply projects/pt1/e2e_testing/xfail_sets.py:2039]: I can take a look after this patch of yours getting merged.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

**Comment by vivekkhandelwal1** [comment]
Hi @vinayakdsci, after a small change your patch is good to go. Also, please mark all those conversations as resolved which are fixed. Also, one more test is passing as can be seen in CI, please remove that from the xfail list. If it's passing only for the stable torch version then you can do something like: https://github.com/llvm/torch-mlir/commit/6e84752c395a828eb612b21be4ab26d9f7e60b22#diff-9e059002448f477a0e7ca75f6d440b9c4d0bfe9b092851afce938fdbe51b7368R2194-R2199

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
Comment (vinayakdsci) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:138]: The CI tests for torch-nightly seems to have been cancelled, so I am unable to judge whether this passes only for stable or for both. Is it OK if I remove it from the file altogether?

**Comment by vinayakdsci** [comment]
Hi @vivekkhandelwal1, the CI for torch-nightly appears to have been cancelled, so I am unable to judge whether the test passes only for stable or for both. Is it OK, if I remove it altogether from the file?

**Comment by vinayakdsci** [comment]
Currently I have added a check against nightly, so stable should pass. Thanks!

**Review by vivekkhandelwal1 — APPROVED** [review]

**Pull request merged**
